packages/opentelemetry-basic-tracer/src/BasicTracer.ts

```diff
@@ -0,0 +1,156 @@
+  randomSpanId,
+  randomTraceId,
+} from '@opentelemetry/core';
+import { BasicTracerConfig } from '../src/types';
+
+/**
+ * This class represents a basic tracer.
+ */
+export class BasicTracer implements types.Tracer {
+  static INVALID_ID = '0';
+
+  private defaultAttributes: types.Attributes;
+  private logger: types.Logger;
+  private sampler: types.Sampler;
+  private scopeManager?: BaseScopeManager;
+
+  // TODO: consume invalid span context from `SpanContext.INVALID`
+  static defaultSpan = new NoopSpan({
+    traceId: BasicTracer.INVALID_ID,
+    // TODO: consume invalid `spanId` from `Span.INVALID`
+    spanId: '0',
+    traceOptions: types.TraceOptions.UNSAMPLED,
+  });
+
+  /**
+   * Constructs a new Tracer instance.
+   */
+  constructor(config: BasicTracerConfig = {}) {
+    this.sampler = config.sampler || ALWAYS_SAMPLER;
+    this.scopeManager = config.scopeManager;
+    this.defaultAttributes = config.defaultAttributes || {};
+    this.logger = config.logger || new NoopLogger();
+  }
+
+  /**
+   * Starts a new Span or returns the default NoopSpan based on the sampling
+   * decision.
+   */
+  startSpan(name: string, options: types.SpanOptions = {}): types.Span {
+    let parentSpanContext: types.SpanContext | undefined;
+
+    // parent is a SpanContext
+    if (
+      options.parent &&
+      typeof (options.parent as types.SpanContext).traceId
+    ) {
+      parentSpanContext = options.parent as types.SpanContext;
+    }
+    // parent is a Span
+    if (
+      options.parent &&
+      typeof (options.parent as types.Span).context === 'function'
+    ) {
+      parentSpanContext = (options.parent as types.Span).context();
+    }
+
+    // make sampling decision
+    if (!this.sampler.shouldSample(parentSpanContext)) {
+      return BasicTracer.defaultSpan;
+    }
+
+    // span context
+    const traceId = parentSpanContext
+      ? parentSpanContext.traceId
+      : randomTraceId();
+    const spanId = randomSpanId();
+
+    // TODO: create a real Span
+    const span = new NoopSpan({
+      traceId,
+      spanId,
+    });
+
+    // Set default attributes
+    span.setAttributes(this.defaultAttributes);
+
+    return span;
+  }
+
+  /**
+   * Returns the current Span from the current context.
+   */
+  getCurrentSpan(): types.Span {
+    // Return with defaultSpan if no scope manager provided.
+    if (!this.scopeManager) {
+      this.logger.warn(
+        'getCurrentSpan() returns an invalid default span without a scopeManager'
+      );
+      return BasicTracer.defaultSpan;
+    }
+
+    // Get the current Span from the context.
+    return this.scopeManager.active() as types.Span;
+  }
+
+  /**
+   * Enters the scope of code where the given Span is in the current context.
+   */
+  withSpan<T extends (...args: unknown[]) => ReturnType<T>>(
+    span: types.Span,
+    fn: T
+  ): ReturnType<T> {
+    if (!this.scopeManager) {
+      this.logger.warn('withSpan(...) has no effect without a scopeManager');
```

Comment:
I don't think there should be a warning for this as it's a perfectly valid use case, especially if you use the low level construct.

Reply:
My thinking was that base tracer has no `ScopeManager` by default so calling ScopeManager related functions like `withSpan` is not valid. As it won't do what you would expect. To suppress the warning the user can pass a `NoopScopeManager`. I think it's okay as in the automatic tracer we will ship OT with `CLSScopeManager` and `ZoneScopeManager`, so this warning is really for people who try to do something custom, and I believe for them it's quite useful.

Reply:
Part of what I'm hoping for from the browser side is code simplicity (as in fewer total bytes of code shipped to the client), so if we can find a way to avoid null checks and warning messages by designing the interface a bit differently I'd prefer that.

What if we had something like `new NoopScopeManager({warn: true})` by default that would write warning logs any time someone attempts to call one of its methods? That way in the browser, where we use something like a `GlobalWindowScopeManager` the code for the if statements and warning logs can get tree shaken out (at least in principle).

Reply:
@hekike 
Since `NoopScopeManager` will have the same behavior as the current one (just calling `fn`, it *almost* cost nothing to use it and avoid null check everywhere. If you think warning user is important then a option like @draffensperger suggested would be the way to go for me.